KAFKA-6761: Reduce Kafka Streams Footprint

bbejeck opened the pull request:
This PR is the first pass on the topology optimization effort. Here's a guide to the changes and the order of the list is meant to imply a possible review order.  

1. `StreamsTopolgyGraphImpl` - This class is used to build the graph, record any metadata to be used later during the topology building process. Some stubbed out code to that end is already present
2. `TopologyOptimizerImpl` - This class traverses the graph and uses information contained in a `StreamGraphNode` to make calls to the `InternalTopologyBuilder` to construct the topology.  It's at this point any optimizations will be applied.
3. `StreamsBuilder` -  The `StreamsBuilder#build` method has been modified to call `InternalStreamdBuilder#buildAndOptimize` to build topology.
4. `InternalStreamsBuilder` - Contains a reference to the `StreamsTopologyGraph` added a methods `InternalStreamsBuilder#addNode`  and `InternalStreamsBuilder#buildAndOptimize` `InternalStreamsBuilder` is the only intermediary for building the graph from the DSL.
5. `KStreamImpl`, `KTableImpl`, `KGroupedStream` and `KGroupedTable` and others have been refactored to create `StreamGraphNode` instances and place topology information in the graph vs. calling `InternalTopologyBuilder`.
6. `StreamsGraphNode` and sub-classes are used to contain information from the DSL calls.  IMHO this is the part of the PR so far that needs the most refactoring work from me, but these are merely container objects so the refactoring should be straightforward.  I plan on updating the PR with updates to these classes soon.

Some other notes:
1. I have not added units test for the new code (will be added soon).  But IMHO the fact that the framework is used to handle all DSL operations, I feel that all of our current tests passing suffice for the moment.
2. Due to depreciation, I have not included the `KStreamsBuilder` in this initial PR.  I can do so, but I think this takes us a step backward regarding separation of DSL and PAPI.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

bbejeck committed:
KSTREAMS-1304: start of topology optimization

bbejeck committed:
KSTREAMS-1304: simplified checking for source node

bbejeck committed:
KSTREAMS-1304: needed to invert boolean

bbejeck committed:
checkstyle fixes

bbejeck commented:
@dguy @guozhangwang @mjsax @vvcephei 

This is WIP PR for the sync meeting tomorrow.  This is meant more for clarifying my direction and will most certainly get an overhaul including tests.

guozhangwang reviewed (commented):
I have a general comment about reusing the existing class v.s. adding new interfaces: currently the `InternalTopologyBuilder.addXXX` functions does not actually construct the ProcessorTopology, but just populates the maps like `stateFactories` and `nodeFactories`. So instead of adding these new interfaces, could we 

1. enhance the capabilities of its classes like NodeFactory to include the additional metadata in `ProcessDetails` and `StreamsGraphNode`, 

2. Change the `addXXX` functions of `InternalTopologyBuilder` to not populate some other maps like `stateStoreNameToSourceTopics` and `storeToChangelogTopic` because they are "final" topology information; instead only populate the node factory and state factory map, where the nodes and states are similar to the `GraphNodes` with all the additional metadata.

3. and add an `optimize` function inside `InternalTopologyBuilder` which will manipulate the maps of node factories and state factories like reducing duplicates, re-linking topics to state stores, etc, as an result of the optimization. This is where "the second pass" will happen.

3. Then in the final `private ProcessorTopology build(nodeGroup)` we still generate the topology, but we make sure that `optimize()` is always called before the actual building logic.


I'm thinking about this mainly because while browsing through this PR, I feel many of these new interfaces have duplicated fields and functions with the existing classes and maybe an overkill.

bbejeck committed:
KSTREAMS-1304: moving calls to InternalTopologyBuilder to Optimizer

bbejeck committed:
added build graph step, collect KTable source

bbejeck committed:
moving more calls to graph

bbejeck committed:
Added suppression for Optimizer, added most methods from KStreamImpl to graph

bbejeck committed:
merging join call into topology optimization

bbejeck committed:
added more cases for parsing topology

bbejeck committed:
Added more graph nodes as necessary, organized the StreamsGraph and InternalStreamsBuilder in a better way
some code cleanup

bbejeck commented:
Updated this with some required changes

1. `Moved the StreamTopologyGraphImpl` from `InternalTopolgyBuilder` to `InternalStreamsBuilder`, ` InternalTopologyBuilder` is not involved in optimization process at all other than receiving normal calls when building topology.
2. Some classes have been renamed to something more clear.
2. Added many more intercepts of calls to `InternalToplogyBuilder` to the `StreamsGraphBuilder` to allow for storing data before building topology.
3.  Added classes extending `StreamsGraphNode` I found this unavoidable as some parts of the topology (joins) make several calls to the `InternalTopologyBuilder` which require storing much more additional information.
4. Capturing the calls to the InternalTopologyBuilder is a larger task than estimated.
5. Most of the integration tests are passing but a large number of unit tests still fail.
6. More calls to the `InternalTopologyBuilder` need to be captured and all unit-tests passing before we can move on to the optimization phase.

bbejeck commented:
I'm aware the branch conflicts, but since this is a WIP I'll leave as is for now.

bbejeck commented:
@guozhangwang I understand your comments above, but we need to not change the functioning of the `InternalTopologyBuilder` as it's used in the PAPI.  But right now the approach of capturing all calls to store information is a graph might not be feasible, so I'm trying to think of a hybrid approach that will leave the `InternalTopologyBuilder` intact but allow us to collect the information we need to do the optimizations required.

bbejeck committed:
changes to parsing graph improved number of passing tests

bbejeck committed:
KGroupedTable fixes test improvements down to 31 needing to pass

bbejeck committed:
StreamThread and StreamsBuilder test fixes

bbejeck committed:
StreamsPartitionAssignor QueryableStateIntegration test fixes

bbejeck committed:
fixed KStreamsFineGrainedAutoResetIntegrationTest, AbstractStreamTest,InternalStreamsBuilderTest, KStreamImplTest, assorted KStream and KTable tests.

bbejeck committed:
Fixed StreamsBuilderTest, and updated a KStreamImplTest previously ignored

bbejeck committed:
Fixed StreamsBuilderTest left out of previous commit

bbejeck committed:
clean up combined case statements, added ID to equals use PriorityQueue vs stack or queue

bbejeck committed:
Removed PROCESSING enum and refactored to use PROCESSOR

bbejeck committed:
Cleaned up StreamsTopologyGraphImpl with how predecessors are set corrected AbstractSteamTest

bbejeck committed:
Cleaned up TopologyOptimizerImpl

bbejeck committed:
Fix KStreamBranch nodes for new constructor fixed tests

bbejeck commented:
updated this and rebased.

This is the initial pass on the optimization framework and is ready for review. 

1. All calls to `InternalTopologyBuilder` are now intercepted and used to build a graph.
2. Once the `StreamsBuilder.build()` method is called graph is traversed and used to build topology.
3. At the moment topology is built as is, no optimizations applied.
4. Since `KStreamsBuilder` is deprecated and going away soon, the `KStreamBuilder` has been left out of the optimization process and  `KStreamBuilder` tests skipped.
5. There currently no tests for the new code, as I felt having all of streams tests passing was sufficient for the moment.  However, tests for the new classes will be coming as an update to this PR.
6. The model objects representing graph nodes need some refactoring and will be done as an update to this PR.

bbejeck commented:
\cc @guozhangwang @mjsax @vvcephei

bbejeck committed:
reset log level and clean up

bbejeck commented:
failures unrelated.

retest this please

bbejeck commented:
failures unrelated.

retest this please

dguy commented on streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:
```diff
@@ -515,6 +518,11 @@ public synchronized StreamsBuilder addGlobalStore(final StoreBuilder storeBuilde
      * @return the {@link Topology} that represents the specified processing logic
      */
     public synchronized Topology build() {
+        if (!topologyBuilt) {
```
Do we need this boolean?
Also seems a bit odd that `topologyOptimizer.optimize(..)` doesn't return a `Topology` - i understand it is because of existing code, but is there a way we could make it return a `Topology`

bbejeck replied:
Ack, I'll work on this.

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java:
```diff
@@ -67,8 +67,15 @@ public Long apply(K aggKey, V value, Long aggregate) {
-        builder.internalTopologyBuilder.addProcessor(aggFunctionName, aggregateSupplier, sourceName);
-        builder.internalTopologyBuilder.addStateStore(storeBuilder, aggFunctionName);
+
+        ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(aggregateSupplier)
```
Probably want to suppress unchecked. Also, any reason why this and the next locals aren't `final`? i.e., above two are?

bbejeck replied:
Ack

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsTopologyGraph.java:
```diff
@@ -0,0 +1,93 @@
+ * The builder of the Streams Intermediate DAG used to make
+ * optimizations of a Kafka Streams topology and sub-topologies
+ */
+public abstract class StreamsTopologyGraph {
```
Why an abstract class here? Why not an interface and the single implementation? I don't see this adding any value

bbejeck replied:
I had an interface initially, I can't remember the rationale for changing, so putting it back to an interface.

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsTopologyGraph.java:
```diff
@@ -0,0 +1,93 @@
+     *
+     * @return Map&lt;StreamsGraphNode, Set&lt;StreamGraphNode&gt;&gt;
+     */
+    public Map<StreamsGraphNode, Set<StreamsGraphNode>> getRepartitioningNodeToRepartitioned() {
```
following 3 methods don't seem to be used

bbejeck replied:
during this first phase they aren't but when we do optimizations they will be.

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -31,60 +31,70 @@
-                                          consumed.keyDeserializer(),
-                                          consumed.valueDeserializer(),
-                                          topics.toArray(new String[topics.size()]));
+        ProcessDetails processDetails = ProcessDetails.builder().withConsumed(consumed).withSourceTopics(topics).build();
```
Can we fix the generics or mark as unchecked?

dguy replied:
Also, same as above (and elsewhere), mixture of `final` and not `final` locals. They could all be `final` - i don't really care either way, but consistency would be good

bbejeck replied:
Ack

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -117,16 +135,11 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
                                              final boolean isQueryable,
                                              final String source,
-                                             final String name) {
+                                             final String name,
```
this is for L134 - topic is not used

bbejeck replied:
Ack

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -117,16 +135,11 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
-                                          consumed.valueDeserializer(),
-                                          topic);
-        internalTopologyBuilder.addProcessor(name, processorSupplier, source);
+        processDetailsBuilder.withProcessorSupplier(processorSupplier);
```
unchecked

bbejeck replied:
Ack

dguy reviewed (commented):
Thanks @bbejeck, i haven't managed to review it all but left a few comments to start with

bbejeck committed:
updates per comments - made locals final, suppressed unchecked warnings

bbejeck committed:
updates per comments - made StreamsTopologyGraph interface

bbejeck commented:
@dguy thanks for the comments.  

updated this.

bbejeck committed:
refactored to remove TopologyOptimizer from StreamsBuilder and placed inside InternalStreamsBuilder.

bbejeck commented:
updated with refactoring to pull `TopologyOptimizer` out from `StreamsBuilder` and put in `InternalStreamsBuilder`.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+ * {@link AbstractStream} when building a topology
+ */
+@SuppressWarnings({"deprecation", "unchecked"})
+class ProcessDetails<K, V, S extends StateStore> {
```
I suspect the root cause of the lack of strong typing is we enforce the `StateStore` template `S` here, which to me seems unnecessary: we can just enforce typing of key and value for each process detail, but for any enwrapped state store details (if there is any), we can instead make the  `withXXX` templated, while leaving the declared fields like `Materialized<K, V, ? extends StateStore> materialized`.

bbejeck replied:
Ack.  I'll take care of this with the more extensive refactoring of `StreamsGraphNode` into more specific classes.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java:
```diff
@@ -67,8 +67,15 @@ public Long apply(K aggKey, V value, Long aggregate) {
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(aggregateSupplier)
+            .withStoreBuilder(storeBuilder).withConnectProcessorName(sourceName).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(aggFunctionName,
+                                                          TopologyNodeType.AGGREGATE,
```
nit: fix indention (either align to `aggFunctionName` -- or maybe better move `aggFunctionName` to next line and only indent with 4 spaces)

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+        return consumedInternal.valueSerde();
+    }
+
+    public Materialized<K, V, S> getMaterialized() {
```
This function and its corresponding field looks redundant to me: the only callers of this function is checking for null, and when requesting the reference the other `getMaterializedInternal` is used. We can instead only keep the `getMaterializedInternal` and its corresponding fields.

Ditto for produced and consumed.

bbejeck replied:
Ack will include in refactoring for comments.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsGraphNode.java:
```diff
@@ -0,0 +1,174 @@
+
+
+    protected final TopologyNodeType topologyNodeType;
+    protected StreamsGraphNode predecessor;
```
We have merge/join operators. Those should have multiple predecessors. How does this work?

bbejeck replied:
The term predecessor in this sense refers to the graph node immediately preceding the current node, and doesn't correlate to the `predecessorNames` parameter used in the `InternalTopologyBuilder`. 

Graph nodes encapsulate all required information to needed to make the necessary `InternalTopologyBuilder` calls for the particular operator.

I can see having the same name as a parameter used by the `InternalTopologyBuilder` can be confusing.  Maybe rename to `predecessorNode`?  
WDYT?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsGraphNode.java:
```diff
@@ -0,0 +1,174 @@
+
+    protected final TopologyNodeType topologyNodeType;
+    protected StreamsGraphNode predecessor;
+    protected final Collection<StreamsGraphNode> descendants = new LinkedHashSet<>();
```
nit: we use `predecessor` above. maybe use `successor` here (sounds more aligned to me as a non-native speaker -- feel free to ignore)

bbejeck replied:
Ack.  I like `successor` better.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+                           final Consumed<K, V> consumed,
+                           final Materialized<K, V, S> materialized,
+                           final ProcessorSupplier<K, V> processorSupplier,
+                           final StateStoreSupplier<KeyValueStore<K, V>> storeSupplier,
```
I'd suggest we do not include any deprecated classes, because they may soon be removed in the next release.

bbejeck replied:
Ack.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+    private final String sourceName;
+
+
+    private ProcessDetails(final Collection<String> sourceTopics,
```
More generally: I find this class a bit clumsy, and I'm not sure if its brought-in flexibility will be really useful or not. We are keeping all possible fields together in this class, while in the optimizeImpl we are still just using it for specific purposes like `buildSource`, `buildSink`, etc.

I'd suggest consider refactoring this class together with StreamsGraphNode, and have different types of `GraphNodes` each with only a subset of these fields, some examples: `SourceGN`, `SinkGN`, `StatelessGN`, `StatefulGN`, and maybe fined categorized types like `RepartitionGN extends StatelessGN` etc. Admittedly this is not as "expressive" as this class to be leveraged by the optimizer, but I think that is more clean and easier to reason.

bbejeck replied:
Ack, same comments as above.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java:
```diff
@@ -24,7 +24,7 @@
 import java.util.Collections;
 import java.util.Set;
-
+@SuppressWarnings("unchecked")
```
This is a meta comment: we should think hard on not introducing unnecessary unchecked warnings, because they indicate that we are losing on strong typing.

bbejeck replied:
Ack, during refactoring  I'll remove `unchecked` warnings.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:
```diff
@@ -489,7 +489,7 @@ public V apply(final K aggKey, final V aggOne, final V aggTwo) {
     }
 
-    @SuppressWarnings("deprecation")
+    @SuppressWarnings({"deprecation", "unchecked"})
```
Ditto as before, I think the root cause is we should not enforce `StateStore S` template in `ProcessDetails`. And then here we could call `ProcessDetails.<K, T>builder()` to maintain typing enforcement.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TopologyNodeType.java:
```diff
@@ -0,0 +1,47 @@
+    STREAM_GLOBAL_TABLE_JOIN,
+    AGGREGATE,
+    MERGE,
+    KGROUPED_STREAM
```
Nit: this list seems to be randomly arranged -- can be bring in some sorting/grouping or operators -- makes it easier to read the list. (maybe put a comment about what sorting/grouping is applied)

bbejeck replied:
Ack.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+ * {@link AbstractStream} when building a topology
+ */
+@SuppressWarnings({"deprecation", "unchecked"})
+class ProcessDetails<K, V, S extends StateStore> {
```
nit: what about `OperatorDetails` or `NodeDetails` ?

Why do we have use `S extends StateStore` -- atm, DSL uses only one type of store, but this might change -- wondering if we limit ourselves here (refactoring later on, might be cumbersome). Just a thought.

bbejeck replied:
Ack.  I like `OperatorDetails`.  Also the `S` generic is going to be dropped during the refactoring for comments.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TopologyOptimizerImpl.java:
```diff
@@ -0,0 +1,379 @@
+    }
+
+    private void buildSourceSinkNode(final InternalTopologyBuilder internalTopologyBuilder,
+                                     final SourceSinkNode rgn,
```
We should avoid using any abbreviations.

bbejeck replied:
Ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+    private MaterializedInternal<K, V, S> materializedInternal;
+    private ProducedInternal<K, V> producedInternal;
+    private ConsumedInternal<K, V> consumedInternal;
+    private final StateStoreSupplier<KeyValueStore<K, V>> storeSupplier;
```
In general, an operator can have multiple stores, also of different types. Also not, that we support `Transformers` that might have multiple different stores.

bbejeck replied:
Ack.   Note that in addressing these comments the hierarchy of graph objects will change to incorporate these changes and result in more specific graph nodes to account for the different types including a `StatefulGN` that could possibly have multiple different stores.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+    private ProducedInternal<K, V> producedInternal;
+    private ConsumedInternal<K, V> consumedInternal;
+    private final StateStoreSupplier<KeyValueStore<K, V>> storeSupplier;
+    private final StoreBuilder<KeyValueStore<K, V>> storeBuilder;
```
as above.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SourceSinkNode.java:
```diff
@@ -0,0 +1,187 @@
+        }
+
+        public Builder withFilterName(String filterName) {
+            this.filterName = filterName;
```
I'm not sure I understand what does this `filterName` used for, could you explain? It seems to be just used as the function name.

bbejeck replied:
It represents the name of the `ProcessorSupplier` used in `KStreamImpl.createRepartitionedSource`.  

The `filterName` eventually becomes the `funcOrFilterName` variable on the `SourceSinkNode`.  This name was chosen as in `KStreamImpl.createRepartitionedSource` we use a `filterName` for the `ProcessorSupplier`, but in `KGroupedTableImpl.buildAggregate` (again creating a repartitioned topic) we use `funcName` to name the `ProcessorSupplier`.  

The name chosen it too specific and should be something like `processorSupplierName` and the method on the builder should be `withProcessorSupplierName` or someting along those lines.

Additionally, the `SourceSinkNode` is used represent a repartitioned source, but the name will change to `RepartitionedGraphNode` as suggested.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+    private final ProcessorSupplier<K, V> processorSupplier;
+    private MaterializedInternal<K, V, S> materializedInternal;
+    private ProducedInternal<K, V> producedInternal;
+    private ConsumedInternal<K, V> consumedInternal;
```
why do we need `Consumed`/`Produced`/`Materialized` and `ConsumedInternal`/`ProducedInternal`/`MaterializedInternal` -- are the second ones on subsuming the first ones?

bbejeck replied:
This is from trying to capture all possible values in one object; this will be refactored and pared down to only contain required values in more specific graph nodes.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+    private final String[] storeNames;
+    private final String connectProcessorName;
+    private final String[] parentNames;
+    private final String sourceName;
```
Is a `sourceName` not a `parentName` ?

bbejeck replied:
As above, too many fields contained in one object will be cleaned up in addressing comments.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsGraphNode.java:
```diff
@@ -0,0 +1,174 @@
+ * the actual topology.  This is the base case class and other more specific cases required
+ * more detailed information will extend this class
+ */
+public class StreamsGraphNode {
```
Currently we have the following extensions of this node:

1. KTableJoinGN
2. SourceSinkGN
3. StreamStreamJoinGN

And these classes seem to be used for some specialized optimizations, but won't that restrict the flexibility of the optimizer itself? I suggest we make general extensions of the GN that optimizers can understand and manipulate for. Please see my other comments regarding StreamGN and ProcessDetails.

bbejeck replied:
In this case, I don't think so.  

The `SourceSinkNode` represents repartition graph node, and in addressing these comments, the name will be updated accordingly.

In the case of joins, IMHO we need specific classes to represent these actions, as there is too much information critical for performing these operations to make generalized nodes.  

For example, a Stream Stream join adds the following to the topology via the InternalTopologyBuilder:

1. Two `KStreamJoinWindow` processors
2. Two  `KStreamKStreamJoin` processors
3. One merge processor
4. Two `WindowStore` implementations

Since the Stream-Stream join contains state stores, one could argue this is merely a `StatefulGraphNode`. 

But if you contrast a join to the `KStream.transform` which takes single processor and single state store, I think it's fair to say a join operation requires capturing much more specialized information hence requiring a special object capable of holding such information.   

I could see possible having the various join graph nodes extending a `StatefulGN` or another base object.

WDYT?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+    private final StoreBuilder<KeyValueStore<K, V>> storeBuilder;
+    private final KTableSource<K, V> kTableSource;
+    private final String[] storeNames;
+    private final String connectProcessorName;
```
Not sure what this is?

bbejeck replied:
As above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+
+    public Collection<String> getSourceTopics() {
+        final Set<String> copy = new HashSet<>(sourceTopics);
+        return Collections.unmodifiableSet(copy);
```
Why do we return immutable ? It's already it's own copy.

bbejeck replied:
You're right probably overkill, will remove.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TopologyOptimizerImpl.java:
```diff
@@ -0,0 +1,379 @@
+        }
+    }
+
+    private void buildAndMaybeOptimize(final InternalTopologyBuilder internalTopologyBuilder,
```
Correct me if I'm wrong: my understanding is that the optimization execute by traversing the graph and build based on each single GN. In that case how is that different with our current mechanism? For example, if we want to add an optimization rule to merge multiple repartition topics, how could that rule be executed based on this model??

bbejeck replied:
TLDR; for now we'll have hard-coded rules that will be used to build up different data structures in the `StreamsToplogyGraphImpl`, then the rules are going to be applied once the graph is traversed in the optimizer phase.  An example of what I'm referring to follows here.

While the optimizer will work on one graph node at a time, the key point is decisions for optimizations won't occur isolation.  Under our current model, we add an operation to the topology without any context.  We make one - N number calls to the InternalTopologyBuilder adding all the required parts for the topology, but we lose the contextual information at that point.  While we register the name of the previous operation of a parent operation, this only serves to wire to topology together; there is no context as to what the last action is, or for that matter what operations the DSL added N steps prior.

In this model as we add a node to the graph, the `StreamsTopologyGraphImpl` connects the current node with its predecessor in the `setPredecessor` call, and likewise, the current node is connected with its predecessor as a descendant node (`StreamsTopologyGraphImpl` lines 71 -72).  This connectivity allows for traversing in either direction to inspect the meta-data of a given node enabling for making any optimization decisions.


Secondly, when adding a node to the graph, we can use the type of the node as a hint that there is an opportunity to optimize.  While we won't perform the optimization at this point, we'll record the fact that an operation from the DSL is eligible for optimization. For example, consider the following code:

```
KStream<String, String> inputStream = builder.stream("input-topic");

KStream<String, String> mappedStream = inputStream.map(new KeyValueMapper<String, String, KeyValue<? extends String, ? extends String>>() {
            @Override
            public KeyValue<? extends String, ? extends String> apply(String key, String value) {

                return KeyValue.pair(key.substring(0, 3), value);
            }
        });

mappedStream.groupByKey().windowedBy(TimeWindows.of(5000)).count().toStream().to("count-one-out");
mappedStream.groupByKey().windowedBy(TimeWindows.of(10000)).count().toStream().to("count-two-out");
mappedStream.groupByKey().windowedBy(TimeWindows.of(15000)).count().toStream().to("count-three-out");
```

The node representing the `inputStream.map` gets stored as a key in `Map<StreamsGraphNode, Set<StreamsGraphNode>> repartitioningNodeToRepartitioned` and the child nodes of `mappedStream`, the `groupByKey` nodes, end up stored in the `Set<StreamGraphNode>`.

Now going back to where this fits into the `ToplogyOptimizer` as we traverse the graph to build the final topology, we can check each node to see if there have been any optimizations recorded for this particular node.  It's at that point that we'll apply any of the optimizations.  In this case, we'd check the `repartitioningNodeToRepartitioned` mentioned above, and we would do a manual `through` operation to repartition the `mappedStream` and then update the child nodes of the `groupByKey` to use the new information when added to the topology via the InternalStreamsBuilder.

During this first pass, only building the graph and applying a couple of hard-coded rules are going to be done.  In another pass, we'll develop a mechanism for plugging in different rules.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProcessDetails.java:
```diff
@@ -0,0 +1,341 @@
+        return new Builder<>();
+    }
+
+    public static final class Builder<K, V, S extends StateStore> {
```
Builder for what? Can we improve the name?

bbejeck replied:
Ack.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -31,60 +31,71 @@
+@SuppressWarnings("unchecked")
 public class InternalStreamsBuilder implements InternalNameProvider {
 
     final InternalTopologyBuilder internalTopologyBuilder;
```
Can this be a local variable now? It seems it's not used across methods.

vvcephei replied:
If we alter the `add*StateStore()` methods to put the state stores in the `StreamsTopologyGraph`, this class could actually avoid holding a reference to `InternalTopologyBuilder` at all. It would then become the sole responsibility of the `TopologyOptimizer` to use the `InternalTopologyBuilder`.

In fact, that would also let you drop `buildAndOptimize` and `topologyNeedsBuilding`. `org.apache.kafka.streams.StreamsBuilder#build` would instead just `getTopologyGraph` and pass it to the `TopologyOptimizer`.

Now that I'm looking at this after our chat this morning, It was this reference that made me think these classes were more coupled than they are.

bbejeck replied:
Ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -204,4 +217,20 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
                        stateUpdateSupplier);
     }
+
+    public void addNode(StreamsGraphNode graphNode) {
```
Nit: add `final`

Can this be package private?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -204,4 +217,20 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+        }
+    }
+
+    public StreamsTopologyGraph getTopologyGraph() {
```
nit: this seems to be for testing only -- add comment?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:
```diff
@@ -499,8 +499,17 @@ public V apply(final K aggKey, final V aggOne, final V aggTwo) {
+            .withStoreSupplier(storeSupplier)
+            .withConnectProcessorName(sourceName).build();
+        final StreamsGraphNode streamsGraphNode = new StreamsGraphNode(aggFunctionName,
+                                                                 TopologyNodeType.AGGREGATE,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -127,17 +138,29 @@ public String apply(K key, V value) {
-        builder.internalTopologyBuilder.addProcessor(name, new KStreamFilter<>(predicate, false), this.name);
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(new KStreamFilter<>(predicate, false)).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TopologyNodeType.PROCESSOR,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -127,17 +138,29 @@ public String apply(K key, V value) {
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(new KStreamFilter<>(predicate, false)).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TopologyNodeType.PROCESSOR,
+                                                          this.repartitionRequired,
```
nit: remove `this` if possible -- same below for `this.name`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -150,18 +173,23 @@ public String apply(K key, V value) {
+            }
+        )).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TopologyNodeType.SELECT_KEY,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -170,7 +198,14 @@ public String apply(K key, V value) {
-        builder.internalTopologyBuilder.addProcessor(name, new KStreamMap<>(mapper), this.name);
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(new KStreamMap<>(mapper)).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TopologyNodeType.MAP,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -186,7 +221,14 @@ public String apply(K key, V value) {
-        builder.internalTopologyBuilder.addProcessor(name, new KStreamMapValues<>(mapper), this.name);
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(new KStreamMapValues<>(mapper)).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TopologyNodeType.MAP_VALUES,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -255,7 +297,14 @@ public void print(final Printed<K, V> printed) {
+
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(printedInternal.build(this.name)).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TopologyNodeType.PROCESSOR,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -342,9 +397,16 @@ public void writeAsText(final String filePath,
+
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(new KStreamFlatMapValues<>(mapper)).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TopologyNodeType.PROCESSOR,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -358,22 +420,40 @@ public void writeAsText(final String filePath,
+            final ProcessDetails processDetailsInner = ProcessDetails.builder().withProcessorSupplier(new KStreamPassThrough<K, V>()).build();
+
+            final StreamsGraphNode graphNodeInner = new StreamsGraphNode(childName,
+                                                                   TopologyNodeType.PROCESSOR,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -358,22 +420,40 @@ public void writeAsText(final String filePath,
-        builder.internalTopologyBuilder.addProcessor(branchName, new KStreamBranch(predicates.clone(), childNames), this.name);
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(new KStreamBranch(predicates, childNames)).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(branchName,
+                                                          TopologyNodeType.PROCESSOR,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -358,22 +420,40 @@ public void writeAsText(final String filePath,
-            builder.internalTopologyBuilder.addProcessor(childNames[i], new KStreamPassThrough<K, V>(), branchName);
-            branchChildren[i] = new KStreamImpl<>(builder, childNames[i], sourceNodes, this.repartitionRequired);
+        builder.addNode(graphNode);
+        for (StreamsGraphNode branchChildNode : branchChildNodes) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -394,7 +474,13 @@ public void writeAsText(final String filePath,
-        builder.internalTopologyBuilder.addProcessor(name, new KStreamPassThrough<>(), parentNames);
+        final ProcessDetails processDetails = ProcessDetails.builder().withParentNames(parentNames).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          MERGE,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -421,17 +508,30 @@ public void writeAsText(final String filePath,
-        builder.internalTopologyBuilder.addProcessor(name, new KStreamPeek<>(action, false), this.name);
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(new KStreamPeek<>(action, false)).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          PROCESSOR,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -421,17 +508,30 @@ public void writeAsText(final String filePath,
-        builder.internalTopologyBuilder.addProcessor(name, new KStreamPeek<>(action, true), this.name);
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(new KStreamPeek<>(action, true)).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          PROCESSOR,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -498,29 +598,36 @@ public void to(final String topic, final Produced<K, V> produced) {
+        detailsBuilder.withProduced(produced);
+
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          SINK,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -498,29 +598,36 @@ public void to(final String topic, final Produced<K, V> produced) {
+            .withStoreNames(stateStoreNames)
+            .build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TRANSFORM,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -543,23 +650,39 @@ private void to(final String topic, final ProducedInternal<K, V> produced) {
+            .withStoreNames(stateStoreNames)
+            .build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TRANSFORM_VALUES,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -543,23 +650,39 @@ private void to(final String topic, final ProducedInternal<K, V> produced) {
+            .withStoreNames(stateStoreNames)
+            .build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          PROCESSOR,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -781,7 +908,17 @@ public boolean test(final K1 key, final V1 value) {
+        processDetailsBuilder.withProcessorSupplier(new KStreamGlobalKTableJoin<>(valueGetterSupplier, joiner, keyMapper, leftJoin));
+
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          STREAM_GLOBAL_TABLE_JOIN,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -795,9 +932,19 @@ public boolean test(final K1 key, final V1 value) {
+            .withConnectProcessorName(((KTableImpl<K, ?, V1>) other).name);
+
+        final StreamsGraphNode streamsGraphNode = new StreamsGraphNode(name,
+                                                                 STREAM_KTABLE_JOIN,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -868,6 +1022,14 @@ public boolean test(final K1 key, final V1 value) {
         final SerializedInternal<K, V> serializedInternal = new SerializedInternal<>(serialized);
+
+        final StreamsGraphNode graphNode = new StreamsGraphNode("KGROUPED",
+                                                          KGROUPED_STREAM,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -868,6 +1022,14 @@ public boolean test(final K1 key, final V1 value) {
+                                                          KGROUPED_STREAM,
+                                                          repartitionRequired,
+                                                          ProcessDetails.builder().build(),
+                                                          this.name);
```
nit: remove `this`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -842,7 +989,14 @@ public boolean test(final K1 key, final V1 value) {
+                                                          KGROUPED_STREAM,
+                                                          repartitionRequired,
+                                                          ProcessDetails.builder().build(),
+                                                          this.name);
```
nit: remove `this`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -926,10 +1088,10 @@ public boolean test(final K1 key, final V1 value) {
 
-            KStreamJoinWindow<K1, V1> thisWindowedStream = new KStreamJoinWindow<>(thisWindow.name(),
+            final KStreamJoinWindow<K1, V1> thisWindowedStream = new KStreamJoinWindow<>(thisWindow.name(),
                                                                                    windows.beforeMs + windows.afterMs + 1,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -926,10 +1088,10 @@ public boolean test(final K1 key, final V1 value) {
                                                                                    windows.maintainMs());
-            KStreamJoinWindow<K1, V2> otherWindowedStream = new KStreamJoinWindow<>(otherWindow.name(),
+            final KStreamJoinWindow<K1, V2> otherWindowedStream = new KStreamJoinWindow<>(otherWindow.name(),
                                                                                     windows.beforeMs + windows.afterMs + 1,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -161,14 +171,22 @@ public String queryableStoreName() {
+
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(processorSupplier).withStoreBuilder(builder).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TopologyNodeType.PROCESSOR,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -249,15 +267,21 @@ public String queryableStoreName() {
+
+        final ProcessDetails processDetails = ProcessDetails.builder().withProcessorSupplier(processorSupplier).withStoreSupplier(storeSupplier).build();
+        final StreamsGraphNode graphNode = new StreamsGraphNode(name,
+                                                          TopologyNodeType.PROCESSOR,
```
nit: indention (as above)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -249,15 +267,21 @@ public String queryableStoreName() {
+                                                          TopologyNodeType.PROCESSOR,
+                                                          false,
+                                                          processDetails,
+                                                          this.name);
```
nit: remove `this`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableJoinGraphNode.java:
```diff
@@ -0,0 +1,195 @@
+    final String thisKtableName;
+
+
+    public KTableJoinGraphNode(TopologyNodeType topologyNodeType, ProcessorSupplier joinThisProcessor, ProcessorSupplier joinOtherProcessor,
```
nit: add `final` -- reformat, one parameter per line

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableJoinGraphNode.java:
```diff
@@ -0,0 +1,195 @@
+        private Builder() {
+        }
+
+        public static Builder aKTableJoinGraphNode() {
```
`aKTableJoinGraphNode` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableJoinGraphNode.java:
```diff
@@ -0,0 +1,195 @@
+            return new Builder();
+        }
+
+        public Builder withJoinThisProcessor(ProcessorSupplier joinThisProcessor) {
```
nit: add `final` -- also all other method below

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableJoinGraphNode.java:
```diff
@@ -0,0 +1,195 @@
+    public int hashCode() {
+
+        int
+            result =
```
nit: formatting

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/MaterializedInternal.java:
```diff
@@ -40,6 +40,11 @@ public MaterializedInternal(final Materialized<K, V, S> materialized,
         }
     }
 
+    public MaterializedInternal(Materialized<K, V, S> materialized) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -665,29 +788,33 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
+
+        final String repartitionTopic = baseName + REPARTITION_TOPIC_SUFFIX;
+        final String sinkName = builder.newProcessorName(SINK_NAME);
+        final String filterName = builder.newProcessorName(FILTER_NAME);
```
nit: `nullKeyFilterName`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SourceSinkNode.java:
```diff
@@ -0,0 +1,187 @@
+                          final Serde valueSerde,
+                          final String sinkTopic,
+                          final String sinkName,
+                          final String funcOrFilterName,
```
nit: `nullKeyFilterName`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamStreamJoinGraphNode.java:
```diff
@@ -0,0 +1,234 @@
+        if (!super.equals(o)) {
+            return false;
+        }
+        StreamStreamJoinGraphNode that = (StreamStreamJoinGraphNode) o;
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamStreamJoinGraphNode.java:
```diff
@@ -0,0 +1,234 @@
+
+    @Override
+    public int hashCode() {
+
```
nit: remove empty line

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsGraphNode.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+
+    public void setPredecessor(StreamsGraphNode predecessor) {
```
nit: add `final` -- also other methods below

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsGraphNode.java:
```diff
@@ -0,0 +1,174 @@
+        this.predecessorName = predecessorName;
+    }
+
+    public StreamsGraphNode getPredecessor() {
```
we usually omit the "get" and only call it `predecessor()` -- applies to other methods, too

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/KStreamBuilderTest.java:
```diff
@@ -50,6 +51,7 @@
 import static org.junit.Assert.assertTrue;
 
 @SuppressWarnings("deprecation")
+@Ignore
```
Why this?

bbejeck replied:
Since `KStreamsBuilder` is deprecated and going away at some point, I did not include it in the graph building process, hence the tests are ignored as all calls for building the topology go through the graph building process. 

Maybe this needs a deeper discussion offline.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:
```diff
@@ -143,15 +143,16 @@ public void shouldAllowJoinUnmaterializedJoinedKTable() {
         assertThat(topology.stateStores().size(), equalTo(3));
         assertThat(topology.processorConnectedStateStores("KSTREAM-JOIN-0000000010"), equalTo(Collections.singleton("store")));
-        assertThat(topology.processorConnectedStateStores("KTABLE-MERGE-0000000007"), equalTo(Collections.singleton("store")));
+        assertThat(topology.processorConnectedStateStores("KTABLE-MERGE-0000000006"), equalTo(Collections.singleton("store")));
```
This change indicates an backward incompatible change how the build the topology. I think, this must be fixed.

bbejeck replied:
I had changed the structure of the test during development.  I have reverted this to the original, and the test passes.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TopologyOptimizerImpl.java:
```diff
@@ -0,0 +1,379 @@
+                                 final StreamsGraphNode descendant,
+                                 final ProcessDetails processDetails) {
+        Deserializer keyDeserializer = getDeserializer(processDetails.consumedKeySerde());
+        Deserializer valDeserializer = getDeserializer(processDetails.consumedValueSerde());
```
nit: `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TopologyOptimizerImpl.java:
```diff
@@ -0,0 +1,379 @@
+        Serializer keySerializer = getSerializer(processDetails.producedKeySerde());
+        Serializer valSerializer = getSerializer(processDetails.producedValueSerde());
+        final StreamPartitioner partitioner = processDetails.streamPartitioner();
+        String topic = processDetails.getSinkTopic();
```
nit: `final` (3x)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TopologyOptimizerImpl.java:
```diff
@@ -0,0 +1,379 @@
+
+        String topic = processDetails.getSourceTopicArray()[0];
+        Deserializer keyDeserializer = getDeserializer(processDetails.consumedKeySerde());
+        Deserializer valDeserializer = getDeserializer(processDetails.consumedValueSerde());
```
nit: `final`

mjsax reviewed (commented):
I did one initial pass -- did not dig into all details yet (it's a pretty complex PR). Want to understand a couple of things better before I dig deeper (ie, put couple of questions)

Most comments are nits... There are some patterns for those (did not put a comment an all places -- maybe you can fix other places, too)

vvcephei commented on checkstyle/suppressions.xml:
```diff
@@ -51,13 +51,13 @@
 
     <suppress checks="CyclomaticComplexity"
-              files="(ConsumerCoordinator|Fetcher|Sender|KafkaProducer|BufferPool|ConfigDef|RecordAccumulator|KerberosLogin|AbstractRequest|AbstractResponse|Selector|SslFactory|SslTransportLayer|SaslClientAuthenticator|SaslClientCallbackHandler).java"/>
+              files="(ConsumerCoordinator|Fetcher|Sender|KafkaProducer|BufferPool|ConfigDef|RecordAccumulator|KerberosLogin|AbstractRequest|AbstractResponse|Selector|SslFactory|SslTransportLayer|SaslClientAuthenticator|SaslClientCallbackHandler|TopologyOptimizerImpl|StreamsTopologyGraphImpl|JoinGraphNode).java"/>
```
Should we reconsider adding these supressions? These restrictions help prevent us from writing unmaintainable code. Maybe it's a sign that we should break up some of these classes or add utilities to simplify the top-level logic.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsGraphNode.java:
```diff
@@ -0,0 +1,174 @@
+public class StreamsGraphNode {
+
+
+    protected final TopologyNodeType topologyNodeType;
```
Nit: can we make all the fields private and access them through getters? It might be a style thing, so feel free to object. Off the bat, I'm most concerned by non-final, non-private fields.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsGraphNode.java:
```diff
@@ -0,0 +1,174 @@
+    public int hashCode() {
+
+        return Objects.hash(topologyNodeType, name, id);
+    }
```
Can you elaborate on the rationale of these `equals`/`hashcode` overrides?

Since they only contain a few of the properties of this object, there's a possibility of unexpected behavior calling `equal` or adding them to sets and maps. I think the purpose of the override is to support putting these things in sets/maps for some of the optimization logic. If that's the case, another solution would be to create a composite key at that time, and then removing the equals/hashcode override.

This would help to keep this data structure agnostic to the kinds of optimization algorithms that might be applied to it, and also makes the comparison logic more explicit when it is used.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -31,60 +31,71 @@
     }
 
-    @SuppressWarnings("deprecation")
+    @SuppressWarnings({"deprecation", "unchecked"})
```
I actually think neither of these supressions are necessary here. You mentioned before that you'd make a pass and remove the "unchecked" supressions. While you're doing that, can you re-evaluate other supressions that are currently in the code? I think it might be a good practice to remove supressions when possible.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -31,60 +31,71 @@
+        final ProcessDetails processDetails = ProcessDetails.builder().withConsumed(consumed).withSourceTopics(topics).build();
+        final StreamsGraphNode node = new StreamsGraphNode(name, SOURCE, false, processDetails, TOPOLOGY_ROOT);
+
+        topologyGraph.addNode(node);
```
Maybe use the `addNode()` available on this class for consistency? (applies a few times in this file)

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -358,22 +420,40 @@ public void writeAsText(final String filePath,
-        String branchName = builder.newProcessorName(BRANCH_NAME);
-
-        String[] childNames = new String[predicates.length];
+        KStream<K, V>[] branchChildren = (KStream<K, V>[]) Array.newInstance(KStream.class, predicates.length);
```
nit: can be final.

vvcephei replied:
General question: Do we need to use arrays in these builders?

Using the java collections libraries would give us better generics support so we wouldn't need to cast as much. I think these objects are all going to be eligible for GC once the topology is actually built, therefore they aren't going to live during run time, so heap shouldn't be an issue.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -410,6 +496,7 @@ public void writeAsText(final String filePath,
     @Override
     public KStream<K, V> through(final String topic, final Produced<K, V> produced) {
         final ProducedInternal<K, V> producedInternal = new ProducedInternal<>(produced);
+        Objects.requireNonNull(topic, "topic can't be null");
```
Huh, I didn't know this was available on `Objects`. Neat!

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableJoinGraphNode.java:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Is it necessary to override equals and hashcode here? I didn't find a usage of them.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/MaterializedInternal.java:
```diff
@@ -40,6 +40,11 @@ public MaterializedInternal(final Materialized<K, V, S> materialized,
 
+    public MaterializedInternal(Materialized<K, V, S> materialized) {
+        super(materialized);
+        queryable = storeName() != null;
```
It's dangerous to call overridable instance methods during construction. The recommendation is to make `storeName()` final, static, or private, or avoid it in constructors. This applies to the other constructor as well.

bbejeck replied:
Ack, this will be taken out in the refactoring for comments.

vvcephei reviewed (commented):
Partway through the review. Just a few thoughts...

bbejeck commented:
@mjsax I tried to address the larger concerns with explanations.  For the rest of the comments, they will get addressed with refactoring over the next day or so.

bbejeck commented:
closed in preference for https://github.com/apache/kafka/pull/4923
